src/odemis/acq/orsay_milling.py

```diff
@@ -181,37 +184,77 @@ def do_milling(self):
+            # point changes (and it's not 0).
+            while scanner.CurrentMillingPoint.Actual in (prev_milling_pt, "0"):
+                if self.future.cancelled():
+                    raise futures.CancelledError()
+
                 time.sleep(0.1)
                 if time.time() > milling_activation_timeout:
+                    logging.debug("Current milling point = %s, milling state = %s",
+                                  scanner.CurrentMillingPoint.Actual,
+                                  miller.MillingActivationState.Actual)
                     raise TimeoutError("Milling failed to start up after trying for 10 seconds.")
 
             logging.debug("Now milling a rectangle %sm by %sm during %s passes of %s seconds",
-                          obj_to_mill.primaryAxisX*2, obj_to_mill.secondaryAxisLength*2, iteration, duration)
+                          obj_to_mill.primaryAxisX * 2, obj_to_mill.secondaryAxisLength * 2, iteration, duration)
 
-            milling_execution_timeout = time.time() + duration * iteration + 5
+            # Now wait until the milling is done (the miller will set the state target to 0 at the end)
+            while miller.MillingActivationState.Target != '0':
+                time.sleep(0.1)
+                if time.time() > milling_execution_timeout:
+                    raise TimeoutError(f"Milling operation timed out after {time.time() - milling_start_t}s.")
+
+            # Wait until it's actually stopped
             while miller.MillingActivationState.Actual != '0':
                 time.sleep(0.1)
                 if time.time() > milling_execution_timeout:
-                    raise TimeoutError("Milling operation timed out. Please check whether the milling is done correctly.")
-            if self.future.cancelled():
-                raise futures.CancelledError()
+                    raise TimeoutError(f"Milling operation timed out after {time.time() - milling_start_t}s.")
 
-            logging.debug("Milling completed.")
+            milling_dur = time.time() - milling_start_t
+            if milling_dur < duration * iteration:
+                logging.warning("Milling completely only after %s s, while expected %s s",
+                                milling_dur, duration * iteration)
+            else:
+                logging.debug("Milling completed.")
 
-        # make sure the Unsubscribe is always called even if any exception happens
+            if self.future.cancelled():
+                raise futures.CancelledError()
         finally:
-            miller.MillingActivationState.Unsubscribe(self.on_milling_state)
+            # make sure the Unsubscribe is always called even if any exception happens
+            miller.MillingActivationState.Unsubscribe(self.param_logger)
+            scanner.MillingStatus.Unsubscribe(self.param_logger)
+            # scanner.CurrentSubList.Unsubscribe(self.param_logger)
+            # scanner.CurrentMillingPoint.Unsubscribe(self.param_logger)
```

Comment:
Do we need to keep these commented out lines?

Reply:
Yes, because at the top I also left the Subscribe commented out. It's for hard-core debugging. It's very verbose, but can still be handy if we notice the milling process doesn't work properly (yet).